### PR TITLE
[FIX] web_editor, *: prevent dropping unsafe snippets in model fields

### PR DESCRIPTION
**Pull request by bso-odoo**
*: website_form, website_sale

When an unsafe snippet is dropped into a sanitized HTML model field, its
unsafe content gets removed on save.

This commit prevents unsafe snippets from being dropped into sanitized
HTML model fields.
The "Form Builder", "Product Search" and "Product Search Input" blocks
are now prevented from being dropped or moved into sanitized HTML model
fields.

Steps to reproduce:
- Go to a product page
- Drop a "Product Search" snippet into the product-specific section of the
product
- Save
=> The form was removed.

task-2829961

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by bso-odoo** [comment]
@qsm-odoo Is this the kind of approach you had in mind ?
In further versions, I would add a selector about the embed code snippet - or introduce a general class that would be added to the embed code snippet.

**Comment by qsm-odoo** [comment]
@bso-odoo I don't think so, I don't think it is viable to try and detect if a snippet will be sanitized on the JS side :grimacing: Just declare them as problematic for sanitization explicitly. In my mind, it's really a smaller diff which is needed:
A) Change the t-snippet instruction (python): if "t-cannot-sanitize" (or a better name?) is present, then add a special class on the generated thumbnail (not on the snippet)
A') Other possibility instead of A: doing the same as the `data-drop-near`/`data-drop-in` instructions: declaring a d`ata-cannot-sanitize` on an option targeting the right snippets (via `data-selector`) to directly influence what the `all`/`closest` methods return (kinda like the `data-no-check="true"`)
B) Whatever seems best to you, change the JS accordingly to prevent the marked snippets from being dropped in the sanitized dropzones -> and to detect those dropzones, I don't see the right approach in your PR either: if am not mistaken, you seem to exclude `[data-oe-type=html]` ? There are HTML fields which are not sanitized (like the blog post content one), that's probably something to change in qweb rendering (python side) to have a `data-oe-sanitized` on the right elements :wink: 

After that, since A & A' both require an XML update in stable to mark the problematic snippets, you can also do a patch in JS in stable for the existing problematic snippets (which we'll remove in master). And hopefully if we make snippets be declared as records in DB, we'll be able to make all of this more readable :+1: cc @rdeodoo

**Comment by bso-odoo** [comment]
@qsm-odoo I went for (A) because the option's based selection adds elements for each involved options - and I believe we want to make sure we forbid access in this case, even if another harmless option would have allowed the drop target.

**Comment by bso-odoo** [comment]
@qsm-odoo I ended up adding an optional parameter to `_activateInsertionZones` and `_insertDropzone`.
The [WIP] commit is there only to make testing easier on runbot 13.0, and must be dropped before merge.

File: addons/web_editor/models/ir_qweb_fields.py
Comment (qsm-odoo) [inline addons/web_editor/models/ir_qweb_fields.py:17]: I see that the original implementation of this method in base is checking `if not options['inherit_branding'] and not options['translate']:`, should not the translate option be considered here too?

Actually I am wondering if this really needs to be made as an inheritance, maybe we could just put that in base directly (maybe nothing should be in base TBH, I don't think the branding system is useful except for the web_editor so... but since it's in base, well, let's put everything in base?)

Could use @rdeodoo 's opinion on this whole PR :+1: (don't change anything before he comments back)

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:463]: Yep, it's annoying in 13.0 :grimacing: But this seems not robust enough, grep this: `Associate in-page snippets to their name` there is some code which is probably more robust (if it works... did it 5 years ago... :grimacing:) to search a snippet from its classes. From 14.0, you will just be able to use the `data-snippet` attribute on snippets to do this search (we'll change that in forward-port).

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:467]: Don't search the whole DOM to find editor information. The responsibility of knowing information about snippets is in the `SnippetMenu` class. If you need one of those info in the `SnippetEditor` class, then do a `trigger_up`, handled by the `SnippetMenu`.

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:522]: Use `this` instead of `self` for your new lines, will be easier to forward-port... and just better :)

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:529]: `.is('A, A *')` => `.closest('A').length` same everywhere in the PR :wink:

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:951]: Avoid a new parameter in stable, even if optional. There are other ways here, even if hacky. The easiest is probably:

Instead of
```javascript
this._activateInsertionZones($siblings, $selectorChildren, forbidSanitize);
```
Do
```javascript
this._activateInsertionZonesForbidSanitize = forbidSanitize;
this._activateInsertionZones($siblings, $selectorChildren);
delete this._activateInsertionZonesForbidSanitize;
```
That would actually make the diff smaller too I guess.

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1364]: Don't forget you need to update the .pot file for new translatable terms in stable: https://github.com/odoo/odoo/wiki/Translations#adding-new-terms-when-suggesting-fixes-in-stable-versions

But is that sentence ok? I don't think the notion of "sanitized", "model" and "fields" are known by users... and I don't like saying "for security reasons" to users either cc @rdeodoo @pierre-pde

File: addons/website/views/snippets.xml
Comment (qsm-odoo) [inline addons/website/views/snippets.xml:903]: Actually there are sanitized snippets in 13.0 :wink: Forms cannot be added in 13.0 product descriptions for example :wink: We changed that for 14.0, considering forms safe for those https://github.com/odoo/odoo/commit/388c222c6c4bb7e2fe3e67009b248359ae0fd3db

I guess we could technically need your new `t-forbid-sanitize` to not be a boolean but like "all, accepts-forms, etc"... But let's forget about that for the moment I guess. In any case, in 13.0, you should just add your t-forbid-sanitize instruction on the form snippet :+1:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@rdeodoo @pierre-pde I ping you in some comments of my review here :+1:

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (xmo-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:1364]: Agreed, it probably should just say something along the lines of the snippet being incompatible with the field.

File: addons/web_editor/models/ir_qweb.py
Comment (xmo-odoo) [inline addons/web_editor/models/ir_qweb.py:56]: One issue with this is it requires updating the views in stable. Maybe the information could be inferred instead? e.g. run `html_sanitize` on the snippet and check the result?

It would be more expensive tho. But assuming that's for the snippets sidebar (is it?) that might be an acceptable trade. It might be a good idea to cache the "query results" in that case (e.g. have a global map of snippet => sanitization flags)

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:1364]: Maybe, although strange to see this message and still be able to drag & drop in some places on this "object" page. Also, I think @bso-odoo shows that tooltip whatever page you are on? Either need some more code to only show this tooltip on pages with sanitized fields or maybe use a sentence like this one: "For technical reasons, this block may be incompatible with some areas on this page".

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:1364]: EDIT: actually, that tooltip is just totally useless no? Either you can drop the snippet and it is confusing, either you can't and there is a message on the dropzones where you can't. And the message in those dropzone s("For technical reasons, that snippet cannot be dropped in such an area", the message I suggested before) could be better too, maybe: "For technical reasons, this block cannot be dropped here". "Snippet" is not a word to show to users either. (Reminder: @rdeodoo suggested we show those dropzone anyway so that the user does not think it is a bug that some dropzones appear (unsanitized fields) while some does not (sanitized fields).

File: addons/web_editor/models/ir_qweb.py
Comment (qsm-odoo) [reply addons/web_editor/models/ir_qweb.py:56]: @xmo-odoo Answers here: https://github.com/odoo/odoo/pull/89402#issuecomment-1106362854

1) We cannot detect via html_sanitize, the main case we are trying to solve is the embed code snippet... which is just an empty div that the user can fill **during edition**. That "div" has to be *marked* unsanitizable. We could do both though, but not sure it is worth the trouble.

2) As indicated, we'll mark the unsanitizable snippets via JS in stable too. Agreed, that PR should be in draft :wink:

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: I tried to reuse it, but it makes either too many things match of too few - because of attributes such as `s_parallax_is_fixed`.
I therefore rewrote something similar - explicitly excluding the `s_parallax_...` classes.

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: The parameter being forwarded untouched to `_insertDropzone`, I named it `_insertDropzoneForbidSanitize` instead.

File: addons/website/views/snippets.xml
Comment (bso-odoo) [reply addons/website/views/snippets.xml]: Removed `s_title` and `s_hr` wip
Added `t-forbid-sanitize` on form builder and both product search blocks.

File: addons/web_editor/models/ir_qweb.py
Comment (bso-odoo) [reply addons/web_editor/models/ir_qweb.py:56]: Made all snippets containing `<form>` automatically marked as forbidden in sanitized fields.

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (rdeodoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1525]: Probably still a WIP and to be adapted but

image

=> 

image

I am not sure how is the regular dropzone actually blinking in a very light blue-ish background, but use a very light red-ish background instead, not that yellow :p

File: addons/web_editor/models/ir_qweb.py
Comment (rdeodoo) [inline addons/web_editor/models/ir_qweb.py:57]: is this how it is usually done? I hate it :grimacing: 
Just add the space, we don't care if there is an useless trailing space if the `data-oe-forbid-sanitize` is not set

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (rdeodoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1372]: Same here, remove that space IMHO :-)
It looks ugly when reading it, and we don't really care about it since it is part of a DOM only admins sees.

**Review by rdeodoo — COMMENTED** [review]
The global implementation sounds correct to me (not sure if it needed a deep review yet?). UX/UI wise it seems legit too.

File: addons/web_editor/models/ir_qweb.py
Comment (qsm-odoo) [reply addons/web_editor/models/ir_qweb.py:57]: Boh? How is that better than... just not have that space? :thinking:

Comment (rdeodoo) [reply addons/web_editor/models/ir_qweb.py:57]:
```py
"class='text-center%s'%s%s" % {
   ' text-red' if whatever else '',
   ' data-oe-forbid-sanitize="true"' if forbid_sanitize else '',
   " data-something='something'" if xx else "",
}
```
Is barely readable tbh. It is worth doing if we don't want to messes with the DOM for public user but here, no reason?
It is also bug prone, as everything relies on that ugly leading whitespace in every strings :grimacing: 
```py
"class='text-center %s' %s %s" % {
   'text-red' if whatever else '',
   'data-oe-forbid-sanitize="true"' if forbid_sanitize else '',
   "data-something='something'" if xx else "",
}
```

Comment (rdeodoo) [reply addons/web_editor/models/ir_qweb.py:57]: That just doesn't feel like a detail for me TBH (maybe I am a weird person :grimacing: ), I feel like it changes everything visually and reading wise :o

Comment (bso-odoo) [reply addons/web_editor/models/ir_qweb.py]: Went for the "easy to read" way - and will prefer it from now on.
In terms of personal taste, I'd rather never build XML from strings concatenation/substitution ever... but such as change is not going to happen in this project... :stuck_out_tongue:

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: Also did it for the similar block's drag'n'drop .

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:1531]: Replaced `bg-warning` by `bg-danger`, and overloaded it in drop zones to make its alpha 15%.

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (rdeodoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1605]: Looks quite unusual. That's something done that way but to get rid in master? Write a comment please
// Done this way in stable, will be refactored in master

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (rdeodoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1718]: Same here, add a comment that it should be a new param in master but was done that way in stable? Or I am misunderstanding something here?

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (rdeodoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1497]: $els? This can be more than one el, right?

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (rdeodoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1503]: I am not familiar with this part of the code, could you explain to me why do you need a different cache key when forbidden?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thanks for the changes.
Another set of remarks just reading the code without testing

**Review by bso-odoo — COMMENTED** [review]
Made most of the suggested changes, gave explanations for the two other ones.

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thank you Benoit :+1: 

Would be cool to mention those 2 commits in the commit message, as they are related and helpful to get the full history context:
1. Some HTML fields can be sanitized but not sanitizing FORMS https://github.com/odoo/odoo/commit/388c222c6c4bb7e2fe3e67009b248359ae0fd3db
2. s_map is now able to be used in sanitized fields https://github.com/odoo/odoo/commit/c2e9bd0e60014b6a42931cf300e0f89f8cf7c225

This is LGTM (except the small pending remarks), I'll let @qsm-odoo have a final review here if he wants to. Otherwise, I will have a final testing-functional review (which I didn't not stress tested yest).

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:1374]: Indeed, it should be removed in 15.0, not master

File: addons/website_form/views/snippets.xml
Comment (qsm-odoo) [inline addons/website_form/views/snippets.xml:22]: Try not to overflow 80 characters for comments (same for the other ones)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Commit message:

> With [1] and [2], we need a way to mark snippets as being (in)compatible
with sanitization.

- [1] & [2] you mention are not in 13.0, confusing in this 13.0 PR. You can leave it but mention that it's in future version.
- Nitpicking but, actually, I am not sure how [1] and [2] implies that "we need a way to mark snippets as being (in)compatible with sanitization" :wink: I would be more explicit and say something like:
> We need a way to mark snippets as being (in)compatible with sanitization. It cannot be automatic, as, for example, the snippet introduced at [2] contains an iframe but is compatible with sanitization. However, ...

Something like that.

Thank you :+1:

File: addons/web_editor/static/src/js/editor/snippets.editor.js
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: Verified there were not such elements and removed `has` filter.

**Review by qsm-odoo — APPROVED** [review]
> In 15.0 we will add the distinction between full sanitization and
form-tolerant sanitization introduced at [2].

`[2]` you mention has been introduced in 14.0, so everything (JS TODO and form sanitization removed) can be handled for the 14.0 forward-port. I adapted the commit message :+1: 

Thanks for the work :+1: 

@robodoo r+

**Pull request closed without merging**
